File: cpp/tools/queuesuspend.cpp
// queuesuspend - suspend or resume an RPM queue by name.
#include <iostream>
#include <string>

#include "common.hpp"

int main(int argc, char** argv) {
  if (argc != 3) {
    std::cerr << "usage: queuesuspend <queue> <true|false|toggle>\n";
    return 2;
  }
  const std::string queue = argv[1];
  const std::string state = argv[2];
  if (state != "true" && state != "false" && state != "toggle") {
    std::cout << "Invalid Suspend State\n";
    return 1;
  }

  auto r = rpmtools::connectOrExit();
  const auto queues = rpmtools::queueLookup(r->queueListNames());
  auto it = queues.find(queue);
  if (it == queues.end()) {
    std::cout << "Queue name not found.\n";
    return 1;
  }

  // true/false become booleans; toggle stays a string (matches the Python CLI).
  rpm::Json suspend = (state == "toggle") ? rpm::Json("toggle")
                                          : rpm::Json(state == "true");
  std::cout << r->queueModify({{"qid", rpm::Json(std::stoll(it->second))},
                               {"suspend", suspend}}).dump()
            << "\n";
  return 0;
}

File: rpmcli/standalones/queuesuspend.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from argparse import ArgumentParser as ArgParse
import sys

from rpmcli.standalones import connect


if __name__ == '__main__':
  # Connect and validate the configured RPC key.
  r = connect()
  # Parse command line options.
  A = ArgParse(description = "A Utility to Suspend / Resume RPM Queues.")
  A.add_argument("queue", help = "Queue name to control.")
  A.add_argument("state", help = "State - One of (true, false, toggle).")
  options = A.parse_args()
  if options.state not in ('true', 'false', 'toggle'):
    print("Invalid Suspend State")
    sys.exit(1)
  # Create a reverse lookup for Queue names.
  queues = {str(v): k for k, v in r.queue_list_names().items()}
  states = {'true': True,
            'false': False,
            'toggle': 'toggle'}
  if options.queue not in queues:
    print("Queue name not found.")
    sys.exit(1)
  print(r.queue_modify(qid = int(queues[options.queue]), suspend = states[options.state]))
